Doppler.ContactPolicies.Data.Access/Repositories/ContactPoliciesSettings/ContactPoliciesSettingsRepository.cs

```diff
@@ -36,16 +36,79 @@ public async Task<Entities.ContactPoliciesSettings> GetContactPoliciesSettingsAs
 
+        public async Task<Entities.ContactPoliciesSettings> GetBasicContactPoliciesSettingsAsync(string accountName)
+        {
+            using var connection = await _databaseConnectionFactory.GetConnection();
+            const string query =
+                @"select cp.IdUser, cp.Active, cp.Interval [IntervalInDays], cp.Amount [EmailsAmountByInterval], u.Email [AccountName]
+                from [User] u
+                left join [UserShippingLimit] cp on u.IdUser = cp.IdUser and cp.Enabled = 1
+                where u.Email = @Email;";
+
+            var queryParams = new { Email = accountName };
+            var contactPoliciesSettings = await connection.QueryFirstOrDefaultAsync<Entities.ContactPoliciesSettings>(query, queryParams);
+
+            if (contactPoliciesSettings == null)
+            {
+                return null;
+            }
+
+            if (contactPoliciesSettings.IdUser == null)
+            {
+                return contactPoliciesSettings;
+            }
+
+            return contactPoliciesSettings;
+        }
+
+        public async Task<bool> InsertContactPoliciesSettingsAsync(Entities.ContactPoliciesSettings contactPoliciesToInsert)
+        {
+
+            using var connection = await _databaseConnectionFactory.GetConnection();
+
+            using (var transaction = connection.BeginTransaction())
+            {
+                const string updateQuery =
+                    @"update [UserShippingLimit] set Active = @Active, Interval = @IntervalInDays, Amount = @EmailsAmountByInterval where IdUser = @IdUser;";
+
+                var affectedRows = await connection.ExecuteAsync(updateQuery, contactPoliciesToInsert, transaction: transaction);
+                var deletedRows = await connection.ExecuteAsync(
+                @"delete from [SubscribersListXShippingLimit] where IdUser = @IdUser and  IdSubscribersList not in @Ids;",
+                new { IdUser = contactPoliciesToInsert.IdUser, Ids = contactPoliciesToInsert.ExcludedSubscribersLists.Select(x => x.Id) }, transaction: transaction);
+
+                int insertedRows = 0;
+                foreach (var item in contactPoliciesToInsert.ExcludedSubscribersLists)
+                {
+
+                    insertedRows += await connection.ExecuteAsync(
+                                @"if exists (select IdUser,IdSubscribersList from [SubscribersList] where IdUser = @IdUser and IdSubscribersList = @IdSubscribersList)
+                                if not exists (select IdUser,IdSubscribersList from [SubscribersListXShippingLimit] where IdUser = @IdUser and IdSubscribersList = @IdSubscribersList)
+                                insert [SubscribersListXShippingLimit] (IdUser, IdSubscribersList, Active)
+                                values(@IdUser,@IdSubscribersList, @Active)", new
+                                { IdUser = contactPoliciesToInsert.IdUser, IdSubscribersList = item.Id, Active = true }, transaction: transaction);
+                }
```

Comment:
@MarianoHerrera could you help us with this query?

I am not sure, but I think that there should be a better way to do this with a `where` in place of an `if`.

Also, it should be done in a single one insert statement with multiple values, right?

Reply:
This could be achieve creating a temporary table with all the IdUser and IdSubscribersList... Then remove all those witch already are on SubscribersListXShippingLimit and then insert those who doesn't.

Other function is MERGE, also you should create a temporary table and insert all rows who do not match.

Reply:
Mmm, there should be a simpler way, something like _insert where..._

I will think about it.

Reply:
You could do something like

```
INSERT INTO dbo.SubscribesListXShippingLimit (IdUser, IdSubscribersList, Active)
SELECT IdUser, IdSubscribersList, 1
FROM (SELECT @IdUser AS IdUser, @IdSubscribersList AS IdSubscribersList) x
LEFT JOIN dbo.SubscribesListXShippingLimit  sxsl ON sxsl.IdUser = x.IdUser AND sxsl.IdSubscribersList = x.IdSubscribersList
WHERE sxsl.IdUser IS NULL
```

The only problem here is you will run this query for each pair of IdUser and list... so 2 alternatives
1. Create a temporary table and replace "x" subquery with the temprary table
2. Iterates on code and generate something like this as subquery
```
SELECT 1 as IdUser, 2 as IdSubscribersList
UNION 
SELECT 3 as IdUser, 4 as IdSubscribersList
UNION 
SELECT 5 as IdUser, 6 as IdSubscribersList

```

Reply:
@MarianoHerrera do you have an example with `merge`?

Reply:
Oh, do not worry, it seems that using `merge` is not a good idea:

> Performance Tip: The conditional behavior described for the `MERGE` statement works best when the two tables have a complex mixture of matching characteristics. For example, inserting a row if it doesn't exist, or updating a row if it matches. When simply updating one table based on the rows of another table, improve the performance and scalability with basic `INSERT`, `UPDATE`, and `DELETE` statements. For example:
>
> ```sql
> INSERT tbl_A (col, col2)  
> SELECT col, col2
> FROM tbl_B
> WHERE NOT EXISTS (SELECT col FROM tbl_A A2 WHERE A2.col = tbl_B.col);  
> ```
> [MERGE (Transact-SQL)](https://docs.microsoft.com/en-us/sql/t-sql/statements/merge-transact-sql?view=sql-server-ver15#merge-transact-sql)

Reply:
@MarianoHerrera what do you think about compound a query like this:

```sql
DECLARE @userEmail VARCHAR(MAX) SET @userEmail = 'prueba@makingsense.com'
DECLARE @idList1 INT SET @idList1 = 500000
DECLARE @idList2 INT SET @idList2 = 583043
DECLARE @idList3 INT SET @idList3 = 623034

INSERT INTO SubscribersListXShippingLimit (IdUser, IdSubscribersList, Active)
SELECT sl.IdUser, sl.IdSubscribersList, 1 AS Active
FROM SubscribersList sl
JOIN [User] u ON u.IdUser = sl.IdUser
LEFT JOIN SubscribersListXShippingLimit slxsl ON slxsl.IdUser = sl.IdUser AND slxsl.IdSubscribersList = sl.IdSubscribersList
WHERE
    slxsl.IdSubscribersList IS NULL
    AND u.Email = @userEmail
    AND sl.IdSubscribersList in (@idList1, @idList2, @idList3)
```

I am not sure about passing the list ids as parameters or simply hard-coding them in the query. The `LEFT JOIN` could be replaced by a `NOT EXISTS` with a subquery, but I am not sure.

Reply:
@jkortega9008 you can even add another new restriction to the previous query to avoid update values when the user is not allowed to:

```sql
SELECT sl.IdUser, sl.IdSubscribersList, 1 AS Active
FROM SubscribersList sl
JOIN [User] u ON u.IdUser = sl.IdUser
JOIN UserShippingLimit usl ON u.IdUser = usl.IdUser -- Here!
LEFT JOIN SubscribersListXShippingLimit slxsl ON slxsl.IdUser = sl.IdUser AND slxsl.IdSubscribersList = sl.IdSubscribersList
WHERE
    slxsl.IdSubscribersList IS NULL
    AND usl.Enabled = 1 -- Here!
    AND u.Email = @userEmail
    AND sl.IdSubscribersList in (@idList1, @idList2, @idList3)
```

Thoughts?

Reply:
Both ways are almost the same, create a variable for each List and use "IN" or pass it in a table. If the amount of list would be small I prefer the former one, if the amount of list would be large i prefer the table.

> The LEFT JOIN could be replaced by a NOT EXISTS with a subquery, but I am not sure.
Yes, are almost the same, in this case I think it will create the same execution plan.

Reply:
the list of ids would be administrated by the application and would be an array of int I presume, the sql injection is when you do something like "Enter your name" and the value of the input is passed directly to the query.

Why you have your doubt about performance?

If you want we could talk, maybe I'm not seeing something.

Reply:
noted, thanks @MarianoHerrera

Reply:
I did not review the changes in the code but your comment looks great.

I did not know about passing lists as parameters, it is great!

Reply:
I did not review the actual implementation, but we seem to agree on the approach, so, I will mark this conversation as resolved, we can open another conversation later.